addons/sale_timesheet/tests/test_sale_timesheet.py

```diff
@@ -671,3 +671,70 @@ def test_change_customer_and_SOL_after_invoiced_timesheet(self):
         self.assertEqual(self.partner_b, task_so1_timesheet2.partner_id, "The Task's second Timesheet entry should have its partner changed, as it was not invoiced and the Task's partner/customer changed.")
         self.assertEqual(so1_product_global_project_so_line, task_so1_timesheet1.so_line, "The Task's first Timesheet entry should not have changed as it was already invoiced (its so_line should still be equal to the first Sales Order line).")
         self.assertEqual(so2_product_global_project_so_line, task_so1_timesheet2.so_line, "The Task's second Timesheet entry should have it's so_line changed, as the Sales Order Item of the Task changed, and this entry was not invoiced.")
+
+    def test_timesheet_upsell(self):
+        """ Test timesheet upselling and email """
+
+        sale_order = self.env['sale.order'].with_context(mail_notrack=True, mail_create_nolog=True).create({
+            'partner_id': self.partner_a.id,
+            'partner_invoice_id': self.partner_a.id,
+            'partner_shipping_id': self.partner_a.id,
+            'user_id': self.user_employee_company_B.id,
+        })
+        # create SO and confirm it
+        uom_days = self.env.ref('uom.product_uom_day')
+        sale_order_line = self.env['sale.order.line'].create({
+            'order_id': sale_order.id,
+            'name': self.product_order_timesheet3.name,
+            'product_id': self.product_order_timesheet3.id,
```

Comment:
Finally, I see you choose another product and so another configuration than the previous one. 🤔  
But it is okay, this product also generates a task.